File: BlueHeron.CronBuilder/Expression/Expression.vb
' The MIT License (MIT)
' 
' Copyright (c) 2020 Michael Bakker
' 
' Permission is hereby granted, free of charge, to any person obtaining a copy
' of this software and associated documentation files (the "Software"), to deal
' in the Software without restriction, including without limitation the rights
' to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
' copies of the Software, and to permit persons to whom the Software is
' furnished to do so, subject to the following conditions:
' 
' The above copyright notice and this permission notice shall be included in all
' copies or substantial portions of the Software.
' 
' THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
' IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
' FITNESS FOR A PARTICULAR PURPOSE And NONINFRINGEMENT. IN NO EVENT SHALL THE
' AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
' LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
' OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
' SOFTWARE.

Imports BlueHeron.Cron.Localization

''' <summary>
''' Object, that represents a Cron expression.
''' </summary>
<DebuggerDisplay("{Expression}")>
Public NotInheritable Class Expression
	Implements IEquatable(Of Expression)

#Region " Objects and variables "

	Private mDisplay As String
	Private mExpression As String
	Private ReadOnly mHumanizer As IHumanizer

#End Region

#Region " Properties "

	''' <summary>
	''' The collection of <see cref="Parameter"/> objects that are part of the expression.
	''' </summary>
	''' <returns>An array of <see cref="Parameter"/> objects</returns>
	Public ReadOnly Property Parameters As IEnumerable(Of Parameter)

	''' <summary>
	''' Human-readable, localized representation of this expression.
	''' </summary>
	Public ReadOnly Property Display As String
		Get
			If String.IsNullOrEmpty(mDisplay) Then
				mDisplay = mHumanizer.Humanize(Me)
			End If
			Return mDisplay
		End Get
	End Property

	''' <summary>
	''' Returns the textual representation of this expression.
	''' </summary>
	Public ReadOnly Property Expression As String
		Get
			If String.IsNullOrEmpty(mExpression) Then
				mExpression = String.Format(fmtExpression, Parameters(ParameterType.Minute).Value.ToString, Parameters(ParameterType.Hour).Value.ToString, Parameters(ParameterType.Day).Value.ToString, Parameters(ParameterType.Month).Value.ToString, Parameters(ParameterType.DayOfWeek).Value.ToString)
			End If
			Return mExpression
		End Get
	End Property

#End Region

#Region " Public methods and functions "

	''' <summary>
	''' Indicates whether the current <see cref="Expression"/> is equal to the object.
	''' </summary>
	''' <param name="other">The object to compare</param>
	''' <returns>True, if both objects are equal</returns>
	Public Shadows Function Equals(other As Object) As Boolean

		If (other Is Nothing) OrElse (Not TypeOf (other) Is Expression) Then
			Return False
		End If

		Return Equals(DirectCast(other, Expression))

	End Function

	''' <summary>
	''' Indicates whether the current <see cref="Expression"/> is equal to another <see cref="Expression"/>.
	''' </summary>
	''' <param name="other">The <see cref="Expression"/> to compare</param>
	''' <returns>True, if both objects are equal</returns>
	Public Shadows Function Equals(other As Expression) As Boolean Implements IEquatable(Of Expression).Equals

		If other Is Nothing Then
			Return False
		End If
		For i As Integer = 0 To 4
			If Parameters(i).Value <> other.Parameters(i).Value Then
				Return False
			End If
		Next

		Return True

	End Function

	''' <summary>
	''' Equality operator.
	''' </summary>
	''' <param name="left">The left <see cref="Expression"/></param>
	''' <param name="right">The right <see cref="Expression"/></param>
	Public Shared Operator =(left As Expression, right As Expression) As Boolean

		Return left.Equals(right)

	End Operator

	''' <summary>
	''' Inequality operator.
	''' </summary>
	''' <param name="left">The left <see cref="Expression"/></param>
	''' <param name="right">The right <see cref="Expression"/></param>
	Public Shared Operator <>(left As Expression, right As Expression) As Boolean

		Return Not left = right

	End Operator

	''' <summary>
	''' Returns the next date and time at which the schedule that is represented by this expression is matched.
	''' </summary>
	''' <returns>A date and time</returns>
	Public Function [Next]() As Date

		Return FindClosestDate(Date.Now, False)

	End Function

	''' <summary>
	''' Returns the first date and time on or after the given date at which the schedule that is represented by this expression is matched.
	''' </summary>
	''' <param name="datum">The date and time to which the closest match in the future must be found</param>
	''' <returns>A date and time</returns>
	Public Function [Next](datum As Date) As Date

		Return FindClosestDate(datum, False)

	End Function

	''' <summary>
	''' Returns the given number of date and time instances on or after the given date at which the schedule that is represented by this expression is matched.
	''' </summary>
	''' <param name="datum">The date and time from which to start matching future occurrences</param>
	''' <param name="count">The number of matched dates to return</param>
	''' <returns>An <see cref="IEnumerable(Of Date)"/></returns>
	Public Iterator Function [Next](datum As Date, count As Integer) As IEnumerable(Of Date)

		For i As Integer = 1 To Math.Max(1, count)
			datum = FindClosestDate(datum, False)
			Yield datum
			datum = datum.AddMinutes(1) ' move to 1 minute after current match
		Next

	End Function

	''' <summary>
	''' Returns the instances of date and time, falling in the given date range, at which the schedule that is represented by this expression is matched.
	''' The <paramref name="toDate"/> must be áfter the <paramref name="fromDate"/> parameter, else zero matches will be returned.
	''' </summary>
	''' <param name="fromDate">The start date and time from which to start matching future occurrences</param>
	''' <param name="toDate">The end date and time after which to stop matching future occurrences</param>
	''' <returns>An <see cref="IEnumerable(Of Date)"/></returns>
	Public Iterator Function [Next](fromDate As Date, toDate As Date) As IEnumerable(Of Date)

		If fromDate > toDate Then
			Yield Nothing
		End If

		Do While fromDate < toDate
			fromDate = FindClosestDate(fromDate, False)
			Yield fromDate
			fromDate = fromDate.AddMinutes(1) ' move to 1 minute after current match
		Loop

	End Function

	''' <summary>
	''' Returns a boolean, determining whether the current date and time are a match for this schedule.
	''' </summary>
	''' <returns>Boolean, True when the current date and time match the date and time pattern defined by this expression</returns>
	Public Function Poll() As Boolean

		Return Poll(Date.Now)

	End Function

	''' <summary>
	''' Returns a boolean, determining whether the given date and time are a match for this schedule.
	''' </summary>
	''' <returns>Boolean, True when the given date and time match the date and time pattern defined by this expression</returns>
	Public Function Poll(datum As Date) As Boolean
		Dim dtmNext As Date = [Next](datum)

		Return (datum.Year = dtmNext.Year) AndAlso (datum.Month = dtmNext.Month) AndAlso (datum.Day = dtmNext.Day) AndAlso (datum.Hour = dtmNext.Hour) AndAlso (datum.Minute = dtmNext.Minute)

	End Function

	''' <summary>
	''' Returns the previous date and time at which the schedule that is represented by this expression was matched.
	''' </summary>
	''' <returns>A date and time</returns>
	Public Function Previous() As Date

		Return FindClosestDate(Date.Now, True)

	End Function

	''' <summary>
	''' Returns the first date and time on or before the given date at which the schedule that is represented by this expression is matched.
	''' </summary>
	''' <param name="datum">The date and time to which the closest match in the past must be found</param>
	''' <returns>A date and time</returns>
	Public Function Previous(datum As Date) As Date

		Return FindClosestDate(datum, True)

	End Function

	''' <summary>
	''' Returns the given number of date and time instances on or before the given date at which the schedule that is represented by this expression is matched.
	''' </summary>
	''' <param name="datum">The date and time to which the closest match in the past must be found</param>
	''' <param name="count">The number of matched dates to return</param>
	''' <returns>An <see cref="IEnumerable(Of Date)"/></returns>
	Public Iterator Function Previous(datum As Date, count As Integer) As IEnumerable(Of Date)

		For i As Integer = 1 To Math.Max(1, count)
			datum = FindClosestDate(datum, True)
			Yield datum
			datum = datum.AddMinutes(-1) ' move to 1 minute before current match
		Next

	End Function

	''' <summary>
	''' Returns the instances of date and time, falling in the given date range, at which the schedule that is represented by this expression is matched.
	''' The <paramref name="toDate"/> must be befóre the <paramref name="fromDate"/> parameter, else zero matches will be returned.
	''' </summary>
	''' <param name="fromDate">The start date and time from which to start matching past occurrences</param>
	''' <param name="toDate">The end date and time after which to stop matching past occurrences</param>
	''' <returns>An <see cref="IEnumerable(Of Date)"/></returns>
	Public Iterator Function Previous(fromDate As Date, toDate As Date) As IEnumerable(Of Date)

		If toDate > fromDate Then
			Yield Nothing
		End If
		Do While fromDate > toDate
			fromDate = FindClosestDate(fromDate, True)
			Yield fromDate
			fromDate = fromDate.AddMinutes(-1) ' move to 1 minute before current match
		Loop

	End Function

	''' <summary>
	''' Returns the string representation of this expression.
	''' </summary>
	<DebuggerStepThrough()>
	Public Overrides Function ToString() As String

		Return Expression

	End Function

#End Region

#Region " Private methods and functions "

	''' <summary>
	''' Returns the first date and time before or after the given date at which the schedule that is represented by this expression is matched.
	''' </summary>
	''' <param name="datum">The date and time to which the closest match in the past or future must be found</param>
	''' <param name="goBack">If True, the closest match in the past is returned, else the closest match in the future is returned</param>
	''' <returns>A <see cref="Date"/></returns>
	Private Function FindClosestDate(datum As Date, goBack As Boolean) As Date
		Dim carry As Integer = 0 ' remembered value to add to next level (-1, 0 or +1)
		Dim minuteMatched, hourMatched, dayMatched, monthMatched, isMatch As Boolean
		Dim matchedMinute, matchedHour, matchedDay, matchedMonth, matchedYear As Integer

		Do Until isMatch ' start with smallest date/time component and work up
			If Not minuteMatched Then
				matchedMinute = FindClosestValue(Parameters(ParameterType.Minute).Matches, datum.Minute, goBack, carry)
				datum = New Date(datum.Year, datum.Month, datum.Day, datum.Hour, matchedMinute, 0).AddHours(carry)
				carry = 0
				minuteMatched = True
			End If
			If Not hourMatched Then
				matchedHour = FindClosestValue(Parameters(ParameterType.Hour).Matches, datum.Hour, goBack, carry)
				If (matchedHour <> datum.Hour) AndAlso (Not Parameters(ParameterType.Minute).Value.ValueType = ValueType.Number) Then ' must recalculate minute
					minuteMatched = False
				End If
				datum = New Date(datum.Year, datum.Month, datum.Day, matchedHour, If(minuteMatched, matchedMinute, If(goBack, MaximumValue(ParameterType.Minute), MinimumValue(ParameterType.Minute))), 0).AddDays(carry) ' update date to match
				carry = 0
				hourMatched = True
			End If
			If Not dayMatched Then
				matchedDay = FindClosestDay(datum, goBack, carry)
				If (matchedDay <> datum.Day) AndAlso (Not Parameters(ParameterType.Hour).Value.ValueType = ValueType.Number) Then  ' must recalculate minute and hour
					minuteMatched = False
					hourMatched = False
				End If
				datum = New Date(datum.Year, datum.Month, matchedDay, If(hourMatched, matchedHour, If(goBack, MaximumValue(ParameterType.Hour), MinimumValue(ParameterType.Hour))), If(minuteMatched, matchedMinute, If(goBack, MaximumValue(ParameterType.Minute), MinimumValue(ParameterType.Minute))), 0).AddMonths(carry)  ' update date to match
				carry = 0
				dayMatched = True
			End If
			If Not monthMatched Then
				matchedMonth = FindClosestValue(Parameters(ParameterType.Month).Matches, datum.Month, goBack, carry)
				matchedYear = datum.Year + If(carry > 0, 1, If(carry < 0, -1, 0))
				If (matchedMonth <> datum.Month) OrElse (matchedYear <> datum.Year) Then  ' must recalculate minute, hour and day
					minuteMatched = False
					hourMatched = False
					dayMatched = False
				End If
				datum = New Date(matchedYear, matchedMonth, If(dayMatched, matchedDay, If(goBack, MaximumValue(ParameterType.Day), MinimumValue(ParameterType.Day))), If(hourMatched, matchedHour, If(goBack, MaximumValue(ParameterType.Hour), MinimumValue(ParameterType.Hour))), If(minuteMatched, matchedMinute, If(goBack, MaximumValue(ParameterType.Minute), MinimumValue(ParameterType.Minute))), 0)  ' update date to match
				carry = 0
				monthMatched = True
			End If
			isMatch = (minuteMatched And hourMatched And dayMatched And monthMatched)
		Loop

		Return datum

	End Function

	''' <summary>
	''' Finds the closest matching number to the given value for the given pattern of numbers, searching either forward or backward.
	''' If the search moves beyond the beginning (looking backward) or end (looking forward), -1 or +1 respectively is carried to the next level (assuming: minute -> hour -> day -> month -> year).
	''' </summary>
	''' <param name="pattern">A <see cref="List(Of Integer)"/></param>
	''' <param name="value">The value to match</param>
	''' <param name="goBack">If true, search backwards for the closest match</param>
	''' <param name="carry">-1, 0 or +1 to be added to the next level</param>
	''' <returns>The closest matching number</returns>
	Private Function FindClosestValue(pattern As IEnumerable(Of Integer), value As Integer, goBack As Boolean, ByRef carry As Integer) As Integer

		If pattern.Contains(value) Then
			Return value ' exact match
		Else
			If goBack Then
				Dim backwardPattern As IEnumerable(Of Integer) = pattern.Where(Function(v) v < value)
				Dim cnt As Integer = backwardPattern.Count

				If cnt = 0 Then
					carry = -1
					Return pattern(pattern.Count - 1) ' highest value of previous cycle
				Else
					Return backwardPattern(cnt - 1) ' closest smaller value
				End If
			Else
				Dim forwardPattern As IEnumerable(Of Integer) = pattern.Where(Function(v) v > value)

				If forwardPattern.Count = 0 Then
					carry = 1
					Return pattern(0) ' lowest value of next cycle
				Else
					Return forwardPattern(0) ' closest higher value
				End If
			End If
		End If

	End Function

	''' <summary>
	''' Finds the closest matching day to the given date for this <see cref="Cron.Expression"/>, searching either forward or backward.
	''' For each cycle that the search moves beyond the beginning (looking backward) or end (looking forward), -1 or +1 respectively, is added to the number to be carried to the next level (assuming: minute -> hour -> day -> month -> year).
	''' </summary>
	''' <param name="datum">The date to match</param>
	''' <param name="goBack">If true, search backwards for the closest match</param>
	''' <param name="carry"></param>
	''' <returns></returns>
	Private Function FindClosestDay(datum As Date, goBack As Boolean, ByRef carry As Integer) As Integer
		Dim daysInMonth As Integer = Date.DaysInMonth(datum.Year, datum.Month)
		Dim dayPattern As List(Of Integer) = Parameters(ParameterType.Day).Matches.Take(daysInMonth).ToList ' current month may be 28, 29, 30 or 31 days in length
		Dim dowValueType As ValueType = Parameters(ParameterType.DayOfWeek).Value.ValueType

		If dowValueType <> ValueType.Any Then
			Dim dayOfWeekPattern As IEnumerable(Of Integer) = Parameters(ParameterType.DayOfWeek).Matches
			Dim filteredDayPattern As New List(Of Integer)

			dayPattern.ForEach(Sub(d)
								   If dayOfWeekPattern.Contains(New Date(datum.Year, datum.Month, d).DayOfWeek) Then
									   filteredDayPattern.Add(d)
								   End If
							   End Sub)

			If dowValueType = ValueType.DayOfWeek Then ' take intersection with DayOfWeek pattern
				dayPattern = filteredDayPattern
			ElseIf dowValueType = ValueType.Symbol_Hash Then ' take intersection with Hash pattern
				Dim n As Integer = Parameters(ParameterType.DayOfWeek).Value.Values(1).Value

				dayPattern = If(n > filteredDayPattern.Count, New List(Of Integer), {filteredDayPattern(n - 1)}.ToList)
			End If
		End If

		If dayPattern.Contains(datum.Day) Then
			Return datum.Day ' exact match
		Else
			If goBack Then
				Dim backwardPattern As IEnumerable(Of Integer) = dayPattern.Where(Function(v) v < datum.Day)
				Dim cnt As Integer = backwardPattern.Count

				If cnt = 0 Then
					carry -= 1
					Return FindClosestDay(New Date(datum.Year, datum.Month, 1).AddDays(-1), goBack, carry) ' continue search starting from the last day of the previous month
				Else
					Return backwardPattern(cnt - 1) ' closest day in past
				End If
			Else
				Dim forwardPattern As IEnumerable(Of Integer) = dayPattern.Where(Function(v) v > datum.Day)

				If forwardPattern.Count = 0 Then
					carry += 1
					Return FindClosestDay(New Date(datum.Year, datum.Month, daysInMonth).AddDays(1), goBack, carry)  ' continue search starting from the first day of the next month
				Else
					Return forwardPattern(0) ' closest day in future
				End If
			End If
		End If

	End Function

#End Region

#Region " Construction "

	''' <summary>
	''' Creates a new <see cref="Expression"/>, using the given parameters.
	''' </summary>
	''' <param name="params">Existing array of <see cref="Parameter"/> objects</param>
	''' <param name="humanizer">The <see cref="IHumanizer"/> to use</param>
	<DebuggerStepThrough()>
	Friend Sub New(params As Parameter(), ByRef humanizer As IHumanizer)

		Parameters = params
		mHumanizer = humanizer

	End Sub

#End Region

End Class

File: BlueHeron.CronBuilder/Expression/Expression.cs
using System.Diagnostics;
using BlueHeron.Cron.Localization;

namespace BlueHeron.Cron;

/// <summary>
/// Object, that represents a Cron expression.
/// </summary>
public sealed class Expression : IEquatable<Expression>
{
    #region Objects and variables

    private string? mDisplay;
    private string? mExpression;
    private readonly IHumanizer mHumanizer;
    private readonly List<Parameter> mParameters;

    #endregion

    #region Construction

    /// <summary>
	/// Creates a new <see cref="Expression"/>, using the given parameters.
	/// </summary>
	/// <param name="params">Existing array of <see cref="Parameter"/> objects</param>
	/// <param name="humanizer">The <see cref="IHumanizer"/> to use</param>
	[DebuggerStepThrough()]
	internal Expression(Parameter[] parameters, IHumanizer humanizer)
    {
        mParameters = [.. parameters];
        mHumanizer = humanizer;
    }

    #endregion

    #region Properties

    /// <summary>
    /// Gets the human-readable, localized representation of this <see cref="Cron.Expression"/>.
    /// </summary>
    public string Display
    {
        get
        {
            mDisplay ??= mHumanizer.Humanize(this);
            return mDisplay;
        }
    }

    /// <summary>
    /// Gets the collection of <see cref="Parameter"/> objects that are part of the expression.
    /// </summary>
    public IReadOnlyList<Parameter> Parameters => mParameters;

    #endregion

    #region Public methods and functions

    /// <summary>
    /// Returns the next date and time that matches the schedule that is represented by this expression.
    /// </summary>
    /// <returns>A <see cref="DateTime"/></returns>
    public DateTime Next()
    {
        return FindClosestDate(DateTime.Now, false);
    }

    /// <summary>
    /// Returns the first date and time on or after the given date and time that matches the schedule that is represented by this expression.
    /// </summary>
    /// <param name="datum">The date and time for which the closest match in the future must be found</param>
    /// <returns>A <see cref="DateTime"/></returns>
    public DateTime Next(DateTime datum)
    {
        return FindClosestDate(datum, false);
    }

    /// <summary>
    /// Returns the given number of date and time instances on or after the given date that match the schedule that is represented by this expression.
    /// </summary>
    /// <param name="datum">The date and time from which to start matching future occurrences</param>
    /// <param name="count">The number of matches to return; should be one or more; lower values result in one match</param>
    /// <returns>An <see cref="IEnumerable{DateTime}"/></returns>
    public IEnumerable<DateTime> Next(DateTime datum, int count)
    {
        count = Math.Max(1, count);
        for (var i = 1; i <= count; i++)
        {
            datum = FindClosestDate(datum, false);
            yield return datum;
            datum = datum.AddMinutes(1); // move to one minute after current match
        }
    }

    /// <summary>
    /// Returns the instances of date and time, falling in the given date range, that match the schedule that is represented by this expression.
    /// The <paramref name="toDate"/> must be áfter the <paramref name="fromDate"/> parameter, else zero matches will be returned.
    /// </summary>
    /// <param name="fromDate">The start date and time from which to start matching future occurrences</param>
    /// <param name="toDate">The end date and time after which to stop matching future occurrences</param>
    /// <returns>An <see cref="IEnumerable{DateTime}"</returns>
    public IEnumerable<DateTime> Next(DateTime fromDate, DateTime toDate)
    {
        if (fromDate > toDate)
        {
            yield break;
        }
        while (fromDate < toDate)
        {
            fromDate = FindClosestDate(fromDate, false);
            yield return fromDate;
            fromDate = fromDate.AddMinutes(1); // move to one minute after current match
        }
    }

    /// <summary>
    /// Returns a boolean, determining whether the current date and time are a match for the schedule represented by this expression.
    /// </summary>
    /// <returns>Boolean, <see langword="true"/> if the current date and time match the date and time pattern defined by this expression; else <see langword="false"/></returns>
    public bool Poll()
    {
        return Poll(DateTime.Now);
    }

    /// <summary>
    /// Returns a boolean, determining whether the givent date and time are a match for the schedule represented by this expression.
    /// </summary>
    /// <param name="datum">The date and time to match</param>
    /// <returns>Boolean, <see langword="true"/> if the given date and time match the date and time pattern defined by this expression; else <see langword="false"/></returns>
    public bool Poll(DateTime datum)
    {
        var dtmNext = Next(datum);
        return (datum.Year == dtmNext.Year) && (datum.Month == dtmNext.Month) && (datum.Day == dtmNext.Day) && (datum.Hour == dtmNext.Hour) && (datum.Minute == dtmNext.Minute);
    }

    /// <summary>
    /// Returns the previous date and time that matches the schedule that is represented by this expression.
    /// </summary>
    /// <returns>A <see cref="DateTime"/></returns>
    public DateTime Previous()
    {
        return FindClosestDate(DateTime.Now, true);
    }

    /// <summary>
    /// Returns the first date and time on or before the given date and time that matches the schedule that is represented by this expression.
    /// </summary>
    /// <param name="datum">The date and time for which the closest match in the past must be found</param>
    /// <returns>A <see cref="DateTime"/></returns>
    public DateTime Previous(DateTime datum)
    {
        return FindClosestDate(datum, true);
    }

    /// <summary>
    /// Returns the given number of date and time instances on or before the given date that match the schedule that is represented by this expression.
    /// </summary>
    /// <param name="datum">The date and time from which to start matching past occurrences</param>
    /// <param name="count">The number of matches to return; should be one or more; lower values result in one match</param>
    /// <returns>An <see cref="IEnumerable{DateTime}"/></returns>
    public IEnumerable<DateTime> Previous(DateTime datum, int count)
    {
        count = Math.Max(1, count);
        for (var i = 1; i <= count; i++)
        {
            datum = FindClosestDate(datum, true);
            yield return datum;
            datum = datum.AddMinutes(-1); // move to one minute before current match
        }
    }

    /// <summary>
    /// Returns the instances of date and time, falling in the given date range, that match the schedule that is represented by this expression.
    /// The <paramref name="toDate"/> must be befóre the <paramref name="fromDate"/> parameter, else zero matches will be returned.
    /// </summary>
    /// <param name="fromDate">The start date and time from which to start matching past occurrences</param>
    /// <param name="toDate">The end date and time after which to stop matching past occurrences</param>
    /// <returns>An <see cref="IEnumerable{DateTime}"</returns>
    public IEnumerable<DateTime> Previous(DateTime fromDate, DateTime toDate)
    {
        if (toDate > fromDate)
        {
            yield break;
        }
        while (fromDate > toDate)
        {
            fromDate = FindClosestDate(fromDate, true);
            yield return fromDate;
            fromDate = fromDate.AddMinutes(-1); // move to one minute before current match
        }
    }

    #endregion

    #region Operators

    /// <summary>
    /// Equality operator.
    /// </summary>
    /// <param name="left">The left <see cref="Expression"/></param>
    /// <param name="right">The right <see cref="Expression"/></param>
    /// <returns><see langword="true"/> if the <see cref="Expression"/>s are equal</returns>
    public static bool operator ==(Expression left, Expression right) => left.Equals(right);

    /// <summary>
    /// Inequality operator.
    /// </summary>
    /// <param name="left">The left <see cref="Expression"/></param>
    /// <param name="right">The right <see cref="Expression"/></param>
    /// <returns><see langword="true"/> if the <see cref="Expression"/>s are not equal</returns>
    public static bool operator !=(Expression left, Expression right) => !left.Equals(right);

    #endregion

    #region Overrides

    /// <summary>
    /// Indicates whether this <see cref="Expression"/> is equal to the other object.
    /// </summary>
    /// <param name="other">The other object</param>
    /// <returns><see langword="true"/> if the objects are equal; else <see langword="false"/></returns>
    public override bool Equals(object? other)
    {
        if (other == null || other.GetType() != typeof(Expression))
        {
            return false;
        }
        return Equals((Expression)other);
    }

    /// <summary>
    /// Indicates whether this <see cref="Expression"/> is equal to the other <see cref="Expression"/>.
    /// </summary>
    /// <param name="other">The other <see cref="Expression"/></param>
    /// <returns><see langword="true"/> if the <see cref="Expression"/>s are equal; else <see langword="false"/></returns>
    public bool Equals(Expression? other)
    {
        if (other is null)
        {
            return false;
        }
        for (var i = 0; i < 5; i++)
        {
            if (mParameters[i].Value != other.mParameters[i].Value)
            {
                return false;
            }
        }
        return true;
    }

    /// <summary>
    /// Returns a hash code, created from the string representation.
    /// </summary>
    public override int GetHashCode()
    {
        return ToString().GetHashCode();
    }

    /// <summary>
    /// Overridden to return the string representation of this <see cref="Expression"/>.
    /// </summary>
    public override string ToString()
    {
        mExpression ??= string.Format(Constants.fmtExpression, mParameters[(int)ParameterType.Minute].Value.ToString(), mParameters[(int)ParameterType.Hour].Value.ToString(), mParameters[(int)ParameterType.Day].Value.ToString(), mParameters[(int)ParameterType.Month].Value.ToString(), mParameters[(int)ParameterType.DayOfWeek].Value.ToString());
        return mExpression;
    }

    #endregion

    #region Private methods and functions

    /// <summary>
    /// Returns the first date and time before or after the given date where the schedule that is represented by this expression is a match.
    /// </summary>
    /// <param name="datum">The date and time for which the closest match in the past or future must be found</param>
    /// <param name="goBack">If <see langword="true"/>, the closest match in the past is returned, else the closest match in the future is returned</param>
    /// <returns>A <see cref="DateTime"/></returns>
    private DateTime FindClosestDate(DateTime datum, bool goBack)
    {
        var carry = 0; // remembered value to add to next level (-1, 0 or +1)
        bool minuteMatched = false, hourMatched = false, dayMatched = false, monthMatched = false, isMatch = false;
        int matchedMinute = 0, matchedHour = 0, matchedDay = 0, matchedMonth, matchedYear;

        while (!isMatch)
        {
            if (!minuteMatched)
            {
                matchedMinute = FindClosestValue(Parameters[(int)ParameterType.Minute].Matches, datum.Minute, goBack, ref carry);
                datum = new DateTime(datum.Year, datum.Month, datum.Day, datum.Hour, matchedMinute, 0).AddHours(carry);
                carry = 0;
                minuteMatched = true;
            }
            if (!hourMatched)
            {
                matchedHour = FindClosestValue(Parameters[(int)ParameterType.Hour].Matches, datum.Hour, goBack, ref carry);
                if (matchedHour != datum.Hour && Parameters[(int)ParameterType.Minute].Value.ValueType != ValueType.Number) // must recalculate minute
                {
                    minuteMatched = false;
                }
                datum = new DateTime(datum.Year, datum.Month, datum.Day, matchedHour, minuteMatched ? matchedMinute : goBack ? Constants.MaximumValue[ParameterType.Minute] : Constants.MinimumValue[ParameterType.Minute], 0).AddDays(carry); // update date to match
                carry = 0;
                hourMatched = true;
            }
            if (!dayMatched)
            {
                matchedDay = FindClosestDay(datum, goBack, ref carry);
                if (matchedDay != datum.Day && Parameters[(int)ParameterType.Hour].Value.ValueType != ValueType.Number) // must recalculate minute and hour
                {
                    minuteMatched = false;
                    hourMatched = false;
                }
                datum = new DateTime(datum.Year, datum.Month, matchedDay, hourMatched ? matchedHour : goBack ? Constants.MaximumValue[ParameterType.Hour] : Constants.MinimumValue[ParameterType.Hour], minuteMatched ? matchedMinute : goBack ? Constants.MaximumValue[ParameterType.Minute] : Constants.MinimumValue[ParameterType.Minute], 0).AddMonths(carry); // update date to match
                carry = 0;
                dayMatched = true;
            }
            if (!monthMatched)
            {
                matchedMonth = FindClosestValue(Parameters[(int)ParameterType.Month].Matches, datum.Month, goBack, ref carry);
                matchedYear = datum.Year + (carry > 0 ? 1 : carry < 0 ? -1 : 0);
                if (matchedMonth != datum.Month || matchedYear != datum.Year) // must recalculate minute, hour and day
                {
                    minuteMatched = false;
                    hourMatched = false;
                    dayMatched = false;
                }
                datum = new DateTime(matchedYear, matchedMonth,
                    dayMatched ? matchedDay : goBack ? Constants.MaximumValue[ParameterType.Day] : Constants.MinimumValue[ParameterType.Day],
                    hourMatched ? matchedHour : goBack ? Constants.MaximumValue[ParameterType.Hour] : Constants.MinimumValue[ParameterType.Hour],
                    minuteMatched ? matchedMinute : goBack ? Constants.MaximumValue[ParameterType.Minute] : Constants.MinimumValue[ParameterType.Minute], 0); // update date to match
                carry = 0;
                monthMatched = true;
            }
            isMatch = minuteMatched & hourMatched & dayMatched & monthMatched;
        }
        return datum;
    }

    /// <summary>
    /// Finds the closest matching day to the given date for this <see cref="Cron.Expression"/>, searching either forward or backward.
    /// For each cycle that the search moves beyond the beginning (looking backward) or end (looking forward), -1 or +1 respectively, is added to the number to be carried to the next level (assuming: minute -> hour -> day -> month -> year).
    /// </summary>
    /// <param name="datum">The date to match</param>
    /// <param name="goBack">If <see langword="true"/>, search backwards in time for the closest match, else search forwards in time</param>
    /// <param name="carry">-1, 0 or +1 to be added to the next level</param>
    /// <returns>The closest matching number</returns>
    private int FindClosestDay(DateTime datum, bool goBack, ref int carry)
    {
        var daysInMonth = DateTime.DaysInMonth(datum.Year, datum.Month);
        var dayPattern = Parameters[(int)ParameterType.Day].Matches.Take(daysInMonth).ToList(); // current month may be 28, 29, 30 or 31 days in length
        var dowValueType = Parameters[(int)ParameterType.DayOfWeek].Value.ValueType;

        if (dowValueType != ValueType.Any)
        {
            var dayOfWeekPattern = Parameters[(int)ParameterType.DayOfWeek].Matches;
            List<int> filteredDayPattern = [];

            dayPattern.ForEach(d =>
            {
                if (dayOfWeekPattern.Contains((int)new DateTime(datum.Year, datum.Month, d).DayOfWeek))
                {
                    filteredDayPattern.Add(d);
                }
            });
            if (dowValueType == ValueType.DayOfWeek) // take intersection with DayOfWeek pattern
            {
                dayPattern = filteredDayPattern;
            }
            else if (dowValueType == ValueType.Symbol_Hash) // take intersection with Hash pattern
            {
                var n = Parameters[(int)ParameterType.DayOfWeek].Value.Values[1].Value;

                dayPattern = n > filteredDayPattern.Count ? [] : [filteredDayPattern[n - 1]] ;
            }
        }
        if (dayPattern.Contains(datum.Day))
        {
            return datum.Day; // exact match
        }
        if (goBack)
        {
            var backwardPattern = dayPattern.Where(v => v < datum.Day).ToList();

            if (backwardPattern.Count == 0)
            {
                carry -= 1;
                return FindClosestDay(new DateTime(datum.Year, datum.Month, 1).AddDays(-1), goBack, ref carry); // continue search starting from the last day of the previous month
            }
            return backwardPattern[^1]; // closest day in past
        }
        else
        {
            var forwardPattern = dayPattern.Where(v => v > datum.Day).ToList();

            if (forwardPattern.Count == 0)
            {
                carry += 1;
                return FindClosestDay(new DateTime(datum.Year, datum.Month, daysInMonth).AddDays(1), goBack, ref carry); // continue search starting from the first day of the next month
            }
            return forwardPattern[0]; // closest day in future
        }
    }

    /// <summary>
    /// Returns the closest matching number to the given value for the given pattern of numbers, searching either forward or backward.
    /// If the search moves beyond the beginning (looking backward) or end (looking forward), -1 or +1 respectively is carried to the next level (assuming: minute -> hour -> day -> month -> year).
    /// </summary>
    /// <param name="pattern">A <see cref="List{int}"/></param>
    /// <param name="value">The value to match</param>
    /// <param name="goBack">If <see langword="true"/>, search backwards in time for the closest match, else search forwards in time</param>
    /// <param name="carry">-1, 0 or +1 to be added to the next level</param>
    /// <returns>The closest matching number</returns>
    private static int FindClosestValue(IReadOnlyList<int> pattern, int value, bool goBack, ref int carry)
    {
        if (pattern.Contains(value))
        {
            return value;
        }
        if (goBack)
        {
            var backwardPattern = pattern.Where(v => v < value).ToList();

            if (backwardPattern.Count == 0)
            {
                carry = -1;
                return pattern[^1]; // highest value of previous cycle
            }
            return backwardPattern[^1]; // closest smaller value
        }
        else
        {
            var forwardPattern = pattern.Where(v => v > value).ToList();

            if (forwardPattern.Count == 0)
            {
                carry = 1;
                return pattern[0]; // lowest value of next cycle
            }
            return forwardPattern[0]; // closest higher value
        }
    }

    #endregion
}